### PR TITLE
Cap the number of stats kept in StatsActor and purge in FIFO order if the limit exceeded

### PR DESCRIPTION
**Pull request by jianoaix**
Signed-off-by: jianoaix <iamjianxiao@gmail.com>

## Why are these changes needed?
There is a risk of using too much of memory in StatsActor, because its lifetime is the same as cluster lifetime.
This puts a cap on how many stats to keep, and purge the stats in FIFO order if this cap is exceeded.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — CHANGES_REQUESTED** [review]
Why choose the more complex logic? There is nothing wrong with creating a admin actor: it consumes zero resources, and you can bound its state with a fifo queue.

The recreate code is very scary and practically guaranteed to break in the future.

**Comment by jianoaix** [comment]
@ericl suggestion adopted. PTAL, thanks.

File: python/ray/data/_internal/stats.py [hunk 1]
```diff
@@ -88,23 +88,40 @@ class _StatsActor:
...
-    The stats data is small so we don't worry about clean up for now.
+    In order to cap memory usage, we set a max number of stats to keep
+    in the actor. When this limit is exceeded, the stats will be garbage
+    collected in FIFO order.
 
     TODO(ekl) we should consider refactoring LazyBlockList so stats can be
     extracted without using an out-of-band actor."""
 
-    def __init__(self):
+    def __init__(self, max_stats=100 * 1000):
```
Comment (ericl) [inline python/ray/data/_internal/stats.py:98]: A bit high?

File: python/ray/data/tests/test_dataset.py [hunk 2]
```diff
@@ -4688,6 +4689,30 @@ def get_node_id():
...
+            )
+        )
+        actor.record_start.remote(uuid)
+        actor.record_task.remote(uuid, task_idx, metadatas[-1])
+    for uuid in range(3):
+        assert ray.get(actor.get.remote(uuid))[0][task_idx] == metadatas[uuid]
+    # Add the fourth stats to exceed the limit.
+    actor.record_start.remote(3)
+    # The first stats (with uuid=0) should have been purged.
+    assert ray.get(actor.get.remote(0))[0] == {}
```
Comment (ericl) [inline python/ray/data/tests/test_dataset.py:4713]: An even stronger test would be to query the internal dict sizes, to verify the deletion actually happened.

File: python/ray/data/_internal/stats.py (hunk 1 above)
Comment (ericl) [reply python/ray/data/_internal/stats.py:98]: How about 1 or 10k?

Comment (jianoaix) [reply python/ray/data/_internal/stats.py]: Reduced to 10k. IIRC there were people talking about number of blocks above 1000, so set it 10k for safety (it won't take significant amount of bytes anyway).

File: python/ray/data/tests/test_dataset.py (hunk 2 above)
Comment (jianoaix) [reply python/ray/data/tests/test_dataset.py:4724]: Added a private method to the StatsActor to query the dict sizes.

File: python/ray/data/_internal/stats.py (hunk 1 above)
Comment (jianoaix) [reply python/ray/data/_internal/stats.py]: Hm it looks uuid is per Dataset not per block, lowering down to 1k.

**Review by ericl — APPROVED** [review]

**Review by clarkzinzow — APPROVED** [review]
Nice!

**Pull request merged**